api/middleware/logger.go

```diff
@@ -0,0 +1,137 @@
+package middleware
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+	jaeger "github.com/uber/jaeger-client-go"
+	macaron "gopkg.in/macaron.v1"
+)
+
+var (
+	logHeaders = false
+)
+
+type LoggingResponseWriter struct {
+	macaron.ResponseWriter
+	errBody []byte // the body in case it is an error
+}
+
+func (rw *LoggingResponseWriter) Write(b []byte) (int, error) {
+	if rw.ResponseWriter.Status() >= 400 {
+		rw.errBody = make([]byte, len(b))
+		copy(rw.errBody, b)
+	}
+	return rw.ResponseWriter.Write(b)
+}
+
+func Logger() macaron.Handler {
+	return func(ctx *Context) {
+		start := time.Now()
+		ctx.Resp = &LoggingResponseWriter{
+			ResponseWriter: ctx.Resp,
+		}
+		rw := ctx.Resp.(*LoggingResponseWriter)
+		ctx.MapTo(ctx.Resp, (*http.ResponseWriter)(nil))
+		ctx.Next()
+
+		// Only log:
+		// - requests that resulted in errors
+		// - requests on /render path
+		if rw.Status() >= 200 && rw.Status() < 300 && !strings.HasPrefix(ctx.Req.URL.Path, "/render") {
+			return
+		}
+
+		content := fmt.Sprintf("ts=%s", time.Now().Format(time.RFC3339Nano))
+
+		traceID, _ := extractTraceID(ctx.Req.Context())
+		if traceID != "" {
+			content += fmt.Sprintf(" traceID=%s", traceID)
+		}
+
+		err := ctx.Req.ParseForm()
+		if err != nil {
+			log.Errorf("Could not parse http request: %v", err)
+		}
+		paramsAsString := ""
+		if len(ctx.Req.Form) > 0 {
+			paramsAsString += "?"
+			paramsAsString += ctx.Req.Form.Encode()
+		}
+		content += fmt.Sprintf(" msg=\"%s %s%s (%v) %v\" orgID=%d", ctx.Req.Method, ctx.Req.URL.Path, paramsAsString, rw.Status(), time.Since(start), ctx.OrgId)
+
+		referer := ctx.Req.Referer()
+		if referer != "" {
+			content += fmt.Sprintf(" referer=%s", referer)
+		}
+		sourceIP := ctx.RemoteAddr()
+		if sourceIP != "" {
+			content += fmt.Sprintf(" sourceIP=\"%s\"", sourceIP)
+		}
+
+		var errorMsg string
+		if rw.Status() < 200 || rw.Status() >= 300 {
+			errorMsg = url.PathEscape(string(rw.errBody))
+		}
+		if errorMsg != "" {
+			content += fmt.Sprintf(" error=\"%s\"", errorMsg)
+		}
+
+		if logHeaders {
+			headers, err := extractHeaders(ctx.Req.Request)
+			if err != nil {
+				log.Errorf("Could not extract request headers: %v", err)
+			}
+			if headers != "" {
+				content += fmt.Sprintf(" headers=\"%s\"", string(headers))
+			}
+		}
+
+		log.Println(colorLog(rw.Status(), content))
+	}
+}
+
+func colorLog(statusCode int, log string) string {
+	if statusCode >= 200 && statusCode <= 202 {
```

Comment:
shouldn't all `2xx` get the same color

Reply:
color? we don't want colored logs?

Reply:
Doesn't hurt to have colors... I don't think the colors would matter much for us, especially not looking at the logs in Loki because I suspect that the colors won't be visible there. But when considering how to increase Metrictank adoption I think colored logs might make a better first impression when somebody looks at MT for the first time.

Reply:
dont add colors. 

Most logging frameworks only use colors when they detect that the logs are being written to a terminal, because only terminals know how to show colors.    If you add the control characters when outputing to destinations that dont support terminal control colors, the control characters are just rendered as text and make the log lines hard to read.

Reply:
I wonder how that's going to look in Loki? is it going to show these control characters? if so then it would be better if there's at least the option to disable the coloring.

Reply:
> Doesn't hurt to have colors... I don't think the colors would matter much for us, especially not looking at the logs in Loki because I suspect that the colors won't be visible there. But when considering how to increase Metrictank adoption I think colored logs might make a better first impression when somebody looks at MT for the first time.

Grafana renders colors when pulling logs from Loki.

Reply:
Removed color logging in d979e98

Reply:
And how does it know what color to use? I would hope it does this by looking at labels or the log level, rather then by looking for terminal control chars. The former seems simpler and better. Typically one filters by log level or keyword anyway, that's what I think we should encourage at least